concrete-integer/src/server_key/crt_parallel.rs

```diff
@@ -0,0 +1,275 @@
+    /// // Encrypt two messages
+    /// let mut ctxt_1 = cks.encrypt_crt(clear_1, basis.clone());
+    /// let mut ctxt_2 = cks.encrypt_crt(clear_2, basis);
+    ///
+    /// // Compute homomorphically a multiplication
+    /// sks.smart_mul_crt_assign_parallelized(&mut ctxt_1, &mut ctxt_2);
+    ///
+    /// // Decrypt
+    /// let res = cks.decrypt_crt(&ctxt_1);
+    /// assert_eq!((clear_1 * clear_2) % 30, res);
+    /// ```
+    pub fn smart_mul_crt_assign_parallelized(
+        &self,
+        ct_left: &mut Ciphertext,
+        ct_right: &mut Ciphertext,
+    ) {
+        ct_left
+            .ct_vec
+            .par_iter_mut()
+            .zip(&mut ct_right.ct_vec)
+            .for_each(|(block_left, block_right)| {
+                self.key.smart_mul_lsb_assign(block_left, block_right);
+            });
+    }
+    pub fn smart_mul_crt_parallelized(
+        &self,
+        ct_left: &mut Ciphertext,
+        ct_right: &mut Ciphertext,
+    ) -> Ciphertext {
+        let mut ct_res = ct_left.clone();
+        self.smart_mul_crt_assign_parallelized(&mut ct_res, ct_right);
+        ct_res
+    }
+
+    /// Computes a PBS for CRT-compliant functions.
+    ///
+    /// # Warning
+    /// This allows to compute programmable bootstrapping over integers under the condition that
+    /// the function is said to be CRT-compliant. This means that the function should be correct
+    /// when evaluated on each modular block independently (e.g. arithmetic functions).
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use concrete_integer::gen_keys;
+    /// use concrete_shortint::parameters::DEFAULT_PARAMETERS;
+    /// let size = 4;
+    ///
+    /// // Generate the client key and the server key:
+    /// let (cks, sks) = gen_keys(&DEFAULT_PARAMETERS, size);
+    ///
+    /// let clear_1 = 28;
+    /// let basis = vec![2, 3, 5];
+    /// // Encrypt two messages
+    /// let mut ctxt_1 = cks.encrypt_crt(clear_1, basis);
+    ///
+    /// // Compute homomorphically the crt-compliant PBS
+    /// sks.pbs_crt_compliant_function_assign_parallelized(&mut ctxt_1, |x| x * x * x);
+    ///
+    /// // Decrypt
+    /// let res = cks.decrypt_crt(&ctxt_1);
+    /// assert_eq!((clear_1 * clear_1 * clear_1) % 30, res);
+    /// ```
+    pub fn pbs_crt_compliant_function_assign_parallelized<F>(&self, ct1: &mut Ciphertext, f: F)
+    where
+        F: Fn(u64) -> u64,
+    {
+        let basis = ct1.message_modulus_vec.clone();
+
+        let accumulators = basis
+            .iter()
+            .cloned()
+            .map(|b| self.key.generate_accumulator(|x| f(x) % b))
+            .collect::<Vec<_>>();
+
+        ct1.ct_vec
+            .par_iter_mut()
+            .zip(&accumulators)
+            .for_each(|(block, acc)| {
+                self.key.keyswitch_programmable_bootstrap_assign(block, acc);
+            });
+    }
+
+    pub fn pbs_crt_compliant_function_parallelized<F>(
+        &self,
+        ct1: &mut Ciphertext,
```

Comment:
ct1 is not needed as `&mut here` since we direcly clone it.

We need to do a cleanup of the concrete-integer API, so I don't know if its 'urgent' to fix this now, as this is reccurent throught the lib

Reply:
i just copied the code from `crt.rs` and replaced `iter` with `par_iter` should i change `crt.rs` as well?

Reply:
Hum I'm not sure what we should do 🤔

Reply:
lets leave it as it is